### PR TITLE
Fuse pad with its producers.

### LINKED ISSUE #2783: Fusion for pad op in Linalg
One option is to have a `linalg.tensor_pad`. If the `linalg.tensor_pad` is the consumer, we can lower it to buffers world by adding the `linalg.fill` op and `subview` op to the top and passing the result of `subview` op to `generic` op as output argument. For example:

```
%0 = linalg.generic ... %arg { ... } (tensor<>) -> tensor<>
%1 = linalg.tensor_pad %0, %init
return %1
```

->

```
linalg.fill %0_buffer, %init
%sub_buffer = subview %0_buffer ...
linalg.generic ... %arg_buffer, %sub_buffer { ... } (memref<>, memref<>) -> ()
```

This only works when the pad is the "last op" (or say there is a buffer for the result tensor of pad op), because we need to make the subview and pass it to generic op.

Haven't get the idea for lowering when a `linalg.tensor_pad` op is a producer. We probably need #2782 feature, so we can have a temp memory in the middle.

Tag @nicolasvasilache for visibility. I will discuss this approach with @nicolasvasilache

### PR DESCRIPTION
**Pull request by MaheshRavishankar**
Fixes #2783

### CODE REVIEW

**Comment by MaheshRavishankar** [comment]
https://reviews.llvm.org/D132720 are the upstream changes needed for this PR.

File: compiler/src/iree/compiler/Tools/init_mlir_dialects.h [hunk 1]
```diff
@@ -67,6 +68,7 @@ inline void registerMlirDialects(DialectRegistry &registry) {
                   shape::ShapeDialect>();
   // clang-format on
   tensor::registerInferTypeOpInterfaceExternalModels(registry);
+  tensor::registerTilingOpInterfaceExternalModels(registry);
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Tools/init_mlir_dialects.h:71]: just found that it should be `registerTilingInterfaceExternalModels`. :p

Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Tools/init_mlir_dialects.h:71]: Haha! Yeah! Haven't touched this branch in a while

**Comment by MaheshRavishankar** [comment]
Currently blocked by #11273

**Comment by MaheshRavishankar** [comment]
@KoolJBlack . Rebased on main... can you share the input model you have.

**Comment by MaheshRavishankar** [comment]
@nicolasvasilache looks like this is in a landable state from within IREE. It gets you what you were looking for with #13042 . Current draft of this PR changes the tiling interface implementation of `tensor.pad` to not generate the `scf.if` (see the change to LLVM associated with this PR here https://github.com/iree-org/iree-llvm-fork/compare/da68d2164efcc1f5e57f090e2ae2219056b120a0...c3b15b0adbf0972bac2c6aae262337a6259214e7) . Unfortunately there is no way to "conditionally load" and interface implementation (its one or the other). It looks like for IREE though we can just avoid generating the `if` conditional during Tiling. It seems to match what you are looking for as well. I think we can change the tiling interface registration to allow for what we need here.
If that is not kosher we can fork the TilingInterface implementation for `tensor.pad` operation in IREE and use the variant that doesnt generate the `if` by default. I am actually leaning towards this solution.

**Comment by nicolasvasilache** [comment]
> Unfortunately there is no way to "conditionally load" and interface implementation (its one or the other). It looks like for IREE though we can just avoid generating the `if` conditional during Tiling. It seems to match what you are looking for as well. I think we can change the tiling interface registration to allow for what we need here.
> If that is not kosher we can fork the TilingInterface implementation for `tensor.pad` operation in IREE and use the variant that doesn't generate the `if` by default. I am actually leaning towards this solution.

I don't grok all the details of the IREE workarounds but the upstream change is incorrect and will potentially miscompile to OOB code for all possible users of the PadOp. "Knowing" that you can take only the else branch because one 1) has done something else before or 2) will do something else after, is an injection of user information.

In particular, note that on GPUs, the assumption you are always making that "tile size is always greater than amount of padding" quickly fails to hold as one distributes the most minor dimension to threadIdx.x with `vector<1xf32>`, `vector<2xf32>` or `vector<4x32>`: the assumption actually almost never holds.

In my case of interest, I have additional information related to 2) that is able to handle this properly.
What works for me and should work for you too, is to provide the user information after tiling with  e.g. https://reviews.llvm.org/D148125.

Can't you just refactor the functionality you need and use it in the particular place you need it?

Also @qcolombet with whom we discussed some of this, and https://github.com/openxla/iree/pull/13042, earlier.

**Comment by MaheshRavishankar** [comment]
> > Unfortunately there is no way to "conditionally load" and interface implementation (its one or the other). It looks like for IREE though we can just avoid generating the `if` conditional during Tiling. It seems to match what you are looking for as well. I think we can change the tiling interface registration to allow for what we need here.
> > If that is not kosher we can fork the TilingInterface implementation for `tensor.pad` operation in IREE and use the variant that doesn't generate the `if` by default. I am actually leaning towards this solution.
> 
> I don't grok all the details of the IREE workarounds but the upstream change is incorrect and will potentially miscompile to OOB code for all possible users of the PadOp. "Knowing" that you can take only the else branch because one 1) has done something else before or 2) will do something else after, is an injection of user information.
> 

There is no workaround here. This is handling pad operations (and fusion with producers and consumers) without having special carve outs for pad op. The issue is the generation of the `if` condition. I know the upstream changes are incorrect (thats why there is a big "Do Not Submit" on the commit message). 

> In particular, note that on GPUs, the assumption you are always making that "tile size is always greater than amount of padding" quickly fails to hold as one distributes the most minor dimension to threadIdx.x with `vector<1xf32>`, `vector<2xf32>` or `vector<4x32>`: the assumption actually almost never holds.
> 
> In my case of interest, I have additional information related to 2) that is able to handle this properly. What works for me and should work for you too, is to provide the user information after tiling with e.g. [reviews.llvm.org/D148125](https://reviews.llvm.org/D148125).
> 
> Can't you just refactor the functionality you need and use it in the particular place you need it?

I dont see why using that is better. It is generating the `if` and removing the `if. I am just not generating the `if` to begin with. Saying that this is user control is a bit of a strange wording. There is no user here, or rather IREE is the user. What I was suggesting is we have an IREE specific implementation of the `TilingInterface` for the `pad` op which basically doesnt even generate the `scf.if` for IREEs. So IREE as a user is injecting this information/taking the burden of making sure that this assertion holds. For example, forking the implementation in IREE will effectively make it unnecessary for you to use that op in IREE. 

Still there are footguns here that makes me uncomfortable to land (I ran the benchmarks here to just check if this assertion holds today... and it seems to, but it is very easy to fall into this hole). I think the `tensor.pad` operation needs to evolve, or using tensor-based codegeneration is hitting limits in terms of abstraction being stretched too much (Uday pointed this out on Discourse too w.r.t to pack operations, and I think he does have a point. At whole program level having tensor-based operations is useful, but within codegen these seem to not play well overall)

**Comment by MaheshRavishankar** [comment]
Superceded by #13133

**Pull request closed without merging**